src/handle.rs

```diff
@@ -432,3 +449,289 @@ impl HandleHost {
+  /// invoked. The invocation is performed solely on a best effort basis. As
+  /// always, GC-based finalization should *not* be relied upon for any critical
+  /// form of resource management!
+  ///
+  /// The callback does not have access to the inner value, because it has
+  /// already been collected by the time it runs.
+  pub fn with_finalizer(
+    isolate: &mut Isolate,
+    handle: impl Handle<Data = T>,
+    finalizer: Box<dyn FnOnce()>,
+  ) -> Self {
+    let HandleInfo { data, host } = handle.get_handle_info();
+    host.assert_match_isolate(isolate);
+    Self::new_raw(isolate, data, Some(finalizer))
+  }
+
+  fn new_raw(
+    isolate: *mut Isolate,
+    data: NonNull<T>,
+    finalizer: Option<Box<dyn FnOnce()>>,
+  ) -> Self {
+    let weak_data = Box::new(WeakData {
+      pointer: Default::default(),
+      finalizer: Cell::new(finalizer),
+    });
+    let data = data.cast().as_ptr();
+    let data = unsafe {
+      v8__Global__NewWeak(
+        isolate,
+        data,
+        weak_data.deref() as *const _ as *const c_void,
+        Self::first_pass_callback,
+      )
+    };
+    weak_data
+      .pointer
+      .set(Some(unsafe { NonNull::new_unchecked(data as *mut _) }));
+    Self {
+      data: Some(weak_data),
+      isolate_handle: unsafe { (*isolate).thread_safe_handle() },
+    }
+  }
+
+  /// Creates a new empty handle, identical to one for an object that has
+  /// already been GC'd.
+  pub fn empty(isolate: &mut Isolate) -> Self {
+    Weak {
+      data: None,
+      isolate_handle: isolate.thread_safe_handle(),
+    }
+  }
+
+  /// Clones this handle and installs a finalizer callback on the clone, as if
+  /// by calling [`Self::with_finalizer`].
+  ///
+  /// Note that if this handle is empty (its value has already been GC'd), the
+  /// finalization callback will never run.
+  pub fn clone_with_finalizer(&self, finalizer: Box<dyn FnOnce()>) -> Self {
+    self.clone_raw(Some(finalizer))
+  }
+
+  fn clone_raw(&self, finalizer: Option<Box<dyn FnOnce()>>) -> Self {
+    if let Some(data) = self.get_pointer() {
+      Self::new_raw(
+        // SAFETY: We're in the isolate's thread, because Weak<T> isn't Send or
+        // Sync.
+        unsafe { self.isolate_handle.get_isolate_ptr() },
+        data,
+        finalizer,
+      )
+    } else {
+      Weak {
+        data: None,
+        isolate_handle: self.isolate_handle.clone(),
+      }
+    }
+  }
+
+  fn get_pointer(&self) -> Option<NonNull<T>> {
+    if let Some(data) = &self.data {
+      // It seems like when the isolate is dropped, even the first pass callback
+      // might not be called.
+      if unsafe { self.isolate_handle.get_isolate_ptr() }.is_null() {
+        None
+      } else {
+        data.pointer.get()
+      }
+    } else {
+      None
+    }
+  }
+
+  pub fn is_empty(&self) -> bool {
+    self.get_pointer().is_none()
+  }
+
+  pub fn open<'a>(&'a self, isolate: &mut Isolate) -> Option<&'a T> {
```

Comment:
I'm not 100% sure this is safe, because the returned pointer is not protected from getting garbage collected. Consider this:

```rust
let f = v8::Weak<v8::Function> = get_fn();
// Called function might trigger garbage collection, which could lead to `f` getting GCed.
f.open().unwrap().call(...); 
```

So maybe it's best to not have this function at all. Users can use `as_local()` instead.

Reply:
Yeah, I'd wondered about that too. Agreed.